mutex.go

```diff
@@ -0,0 +1,20 @@
+package legs
+
+import "sync"
+
+type syncMutex struct {
+	isLocked bool
+	lk       sync.Mutex
+}
+
+func (m *syncMutex) lock() {
+	m.lk.Lock()
+	m.isLocked = true
+}
+
+func (m *syncMutex) unlock() {
+	if m.isLocked {
+		m.lk.Unlock()
+		m.isLocked = false
+	}
+}
```

Comment:
why would we be calling Unlock more than once on a Lock? The panic on doing that is meant to push you to structure your code to not attempt it, rather than make the call to Unlock optional.

And noting that that this construction introduces a potential issue:
in a state where the lock is 'locked', and 2 goroutines, 1 calling lock and one unlock:
* [lock] blocks on the 'Lock()' call
* [unlock] calls 'Unlock'
* [lock] finishes 'Lock()' call. sets `isLocked` to true
* [unlock] sets `isLocked` to false 

you now have a locked mutex, but `isLocked` set to false.

It seems like the ultimate thing you're trying to do is to make sure there's only 1 transfer going at once here, right?
consider maybe a `sync.Once` that's called on error/completion as a way to set the state back to a world where the next action is allowed?

Reply:
I wasn't aware of the existence of `sync.Once`, thank you for the note. The problem with the double unlock was that by exposing `cancelFunc` we take the risk of someone calling `cancelFunc()` and thus unlocking the lock that has been already unlocked when the transfer finished (the tests consider that case, so in the future we should see it blow up if it happens). I think I fixed it now.

And you are right, my mutex construction had parallelization issues I didn't consider, I would have needed to improve/test it to prevent the kind of problems you mentioned.